### PR TITLE
Makefile: initial commit

### PR DESCRIPTION
**Pull request by gyuho**
For https://github.com/coreos/etcd/issues/8502.

First step to make DNS/SRV tests easier to run.

### CODE REVIEW

**Comment by gyuho** [comment]
@xiang90 @fanminshi Please review. This is an initial step to test etcd with DNS/SRV.

Also `Makefile` is needed to make test commands more consistent between different releases.
For every minor release, we just update the `GO_VERSION` string while keeping the same test command `make docker-test`.

Currently, we hard-code Go version for all branches, so it is impossible to test with correct Go versions.

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,145 @@
+.PHONY: build
```
Comment (fanminshi) [inline Makefile]: keep this `# run makefile from repo root`?

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+# run DNS tests inside container
+docker-dns-test-run:
+	docker run \
+	  --rm \
+	  --tty \
+	  --dns 127.0.0.1 \
+	  --volume=`pwd`/bin:/etcd \
+	  --volume=`pwd`/integration/fixtures:/certs \
+	  gcr.io/etcd-development/etcd-dns-test:$(_GO_VERSION) \
+	  /bin/bash -c "pushd /etcd && /run.sh && rm -rf m*.etcd && popd"
```
Comment (fanminshi) [inline Makefile]: maybe use cd instead of pushd?

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+docker-test-pull:
+	docker pull gcr.io/etcd-development/etcd-test:$(_GO_VERSION)
+
+# compile etcd and etcdctl with Linux
+docker-test-compile:
+	docker run \
+	  --rm \
+	  --volume=`pwd`/:/etcd \
+	  gcr.io/etcd-development/etcd-test:$(_GO_VERSION) \
+	  /bin/bash -c "pushd /etcd && GO_BUILD_FLAGS=-v ./build && ./bin/etcd --version"
```
Comment (fanminshi) [inline Makefile]: s/pushd/cd/ ?

**Comment by fanminshi** [comment]
lgtm after nits

**Comment by jpbetz** [comment]
@gyuho Having some trouble with this docker build:

```
make docker-dns-srv-test-build
docker build \
	  --tag gcr.io/etcd-development/etcd-dns-srv-test:go1.9.1 \
	  --file ./e2e/docker-dns-srv/Dockerfile \
	  ./e2e/docker-dns-srv
...
Successfully built 46d8c5294e3f
docker run \
	  --rm \
	  --dns 127.0.0.1 \
	  gcr.io/etcd-development/etcd-dns-srv-test:go1.9.1 \
	  /bin/bash -c "/etc/init.d/bind9 start && cat /dev/null >/etc/hosts && dig +noall +answer SRV _etcd-client-ssl._tcp.etcd.local && dig +noall +answer SRV _etcd-server-ssl._tcp.etcd.local && dig +noall +answer m1.etcd.local m2.etcd.local m3.etcd.local"
WARNING: Localhost DNS setting (--dns=127.0.0.1) may fail in containers.
Starting domain name service...: bind9 failed!
make: *** [docker-dns-srv-test-build] Error 1
```
Still working on getting more detailed logging from a bind9 init.d start failure.  What version of docker where you running this with?

**Comment by gyuho** [comment]
@jpbetz Hmm works for me?

```
Docker version 17.09.0-ce, build afdb6d4

go env
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/gyuho/go"
GORACE=""
GOROOT="/usr/local/go"
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="clang"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/1x/0sf85stn6sb01sx6ncs7x_ph0000gn/T/go-build867045777=/tmp/go-build -gno-record-gcc-switches -fno-common"
CXX="clang++"
CGO_ENABLED="1"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
```

**Comment by jpbetz** [comment]
@gyuho What version of bind9 did docker install for you?
```
$ docker run gcr.io/etcd-development/etcd-dns-srv-test:go1.9.1 apt show bind9 2>/dev/null
Package: bind9
Version: 1:9.10.3.dfsg.P4-12.3+deb9u3
...
```

**Comment by gyuho** [comment]
Looks same?

```
docker run gcr.io/etcd-development/etcd-dns-srv-test:go1.9.1 apt show bind9 2>/dev/null
Package: bind9
Version: 1:9.10.3.dfsg.P4-12.3+deb9u3
...
```
?

Maybe `docker` issue...? Have you tried running with `sudo` if you run on Linux?

**Comment by gyuho** [comment]
Works in my GCP VM too.

```
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.3 LTS
Release:	16.04
Codename:	xenial

$ docker --version
Docker version 17.09.0-ce, build afdb6d4

$ uname -a
Linux gyuho-fast 4.10.0-35-generic #39~16.04.1-Ubuntu SMP Wed Sep 13 09:02:42 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by jpbetz** [comment]
@gyuho It's something with that older docker version. Networking maybe? Everything works with a more recent one. I'm all set now.  Thanks!

**Comment by xiang90** [comment]
@gyuho @jpbetz 

probably the build script should check docker version then? or it would confuse other people too :P.

**Comment by jpbetz** [comment]
@xiang90 Absolutely. I've figured out the incompatibility and submitted a fix (#8661) for the docker files.

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+# keep in-sync with 'Dockerfile-test', 'e2e/docker-dns/Dockerfile'
+_GO_VERSION = go1.9.1
+ifdef GO_VERSION
+	_GO_VERSION = $(GO_VERSION)
+endif
+
+# build base container image for testing on Linux
+docker-test-build:
+	docker build --tag gcr.io/etcd-development/etcd-test:$(_GO_VERSION) --file ./Dockerfile-test .
```
Comment (glevand) [inline Makefile]: You'll need to have some way to distinguish the docker image architecture.  Maybe something like `gcr.io/etcd-development/etcd-test:$(_GO_VERSION)-$(ARCH)`., and also have arch specific docker files.

**Pull request merged**